File: Classes/Thumbs/PDFKThumbCache.m
//
//	ReaderThumbCache.m
//	Reader v2.6.0
//
//	Created by Julius Oklamcak on 2011-09-01.
//	Copyright © 2011-2013 Julius Oklamcak. All rights reserved.
//
//	Permission is hereby granted, free of charge, to any person obtaining a copy
//	of this software and associated documentation files (the "Software"), to deal
//	in the Software without restriction, including without limitation the rights to
//	use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
//	of the Software, and to permit persons to whom the Software is furnished to
//	do so, subject to the following conditions:
//
//	The above copyright notice and this permission notice shall be included in all
//	copies or substantial portions of the Software.
//
//	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
//	OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//	WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
//	CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

#import "PDFKThumbCache.h"
#import "PDFKThumbQueue.h"
#import "PDFKThumbFetcher.h"
#import "PDFKThumbView.h"
#import "PDFKThumbRequest.h"

//The total cost of the cache. 10MB
#define CACHE_SIZE 10485760

@implementation PDFKThumbCache
{
	NSCache *thumbCache;
}

+ (PDFKThumbCache *)sharedCache
{
	static dispatch_once_t onceToken;
    static PDFKThumbCache *cache;
    dispatch_once(&onceToken, ^{
        cache = [self new];
    });
    return cache;
}

+ (NSString *)appCachesPath
{
    // Save a copy of the application caches path the first time it is needed
	static dispatch_once_t predicate = 0;
	static NSString *theCachesPath = nil;
	dispatch_once(&predicate, ^{
        NSArray *cachesPaths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
        theCachesPath = [[cachesPaths objectAtIndex:0] stringByAppendingPathComponent:@"PDFKCache"];
    });
	return theCachesPath;
}

+ (NSString *)thumbCachePathForGUID:(NSString *)guid
{
    //The cache for the PDF document with the given GUID
	NSString *cachesPath = [PDFKThumbCache appCachesPath];
	return [cachesPath stringByAppendingPathComponent:guid];
}

+ (void)createThumbCacheWithGUID:(NSString *)guid
{
	NSFileManager *fileManager = [NSFileManager new];
	NSString *cachePath = [PDFKThumbCache thumbCachePathForGUID:guid];
	[fileManager createDirectoryAtPath:cachePath withIntermediateDirectories:NO attributes:nil error:NULL];
}

+ (void)removeThumbCacheWithGUID:(NSString *)guid
{
	dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
        NSFileManager *fileManager = [NSFileManager new];
        NSString *cachePath = [PDFKThumbCache thumbCachePathForGUID:guid];
        [fileManager removeItemAtPath:cachePath error:NULL];
    });
}

+ (void)touchThumbCacheWithGUID:(NSString *)guid
{
	NSFileManager *fileManager = [NSFileManager new];
	NSString *cachePath = [PDFKThumbCache thumbCachePathForGUID:guid];
	NSDictionary *attributes = [NSDictionary dictionaryWithObject:[NSDate date] forKey:NSFileModificationDate];
	[fileManager setAttributes:attributes ofItemAtPath:cachePath error:NULL];
}

+ (void)purgeThumbCachesOlderThan:(NSTimeInterval)age
{
	dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
        NSDate *now = [NSDate date];
        //Get the list of all caches
        NSString *cachesPath = [PDFKThumbCache appCachesPath];
        NSFileManager *fileManager = [NSFileManager new];
        NSArray *cachesList = [fileManager contentsOfDirectoryAtPath:cachesPath error:NULL];
        
        if (cachesList != nil) {
            for (NSString *cacheName in cachesList) {
                //Get the cache attributes
                NSString *cachePath = [cachesPath stringByAppendingPathComponent:cacheName];
                NSDictionary *attributes = [fileManager attributesOfItemAtPath:cachePath error:NULL];
                //Age of the cache
                NSDate *cacheDate = [attributes objectForKey:NSFileModificationDate];
                NSTimeInterval seconds = [now timeIntervalSinceDate:cacheDate];
                //If older than the age, remove
                if (seconds > age) {
                    [fileManager removeItemAtPath:cachePath error:NULL];
                    #ifdef DEBUG
                        NSLog(@"%s purged %@", __FUNCTION__, cacheName);
                    #endif
                }
            }
        }
    });
}

#pragma mark ReaderThumbCache instance methods

- (id)init
{
	if ((self = [super init]))
	{
		thumbCache = [NSCache new]; // Cache
		[thumbCache setName:@"PDFKThumbCache"];
		[thumbCache setTotalCostLimit:CACHE_SIZE];
	}
	return self;
}

- (id)thumbRequest:(PDFKThumbRequest *)request priority:(BOOL)priority
{
    //We only want one "Run" of this code running at a time. If multiple threads are running this code, we could recreate a thumb twice.
	@synchronized(thumbCache)
	{
        //See if the object exists in the cache.
		id object = [thumbCache objectForKey:request.cacheKey];
        
        //Thumb object does not yet exist in the cache, lets create it.
		if (object == nil)
		{
            //Return an NSNull thumb placeholder object
			object = [NSNull null];
			[thumbCache setObject:object forKey:request.cacheKey cost:2];
            
            //Create a fetching operation
			PDFKThumbFetcher *thumbFetch = [[PDFKThumbFetcher alloc] initWithRequest:request];
            //Set the priority
            [thumbFetch setQueuePriority:(priority ? NSOperationQueuePriorityNormal : NSOperationQueuePriorityLow)];
            request.thumbView.operation = thumbFetch;
            thumbFetch.qualityOfService = NSQualityOfServiceUtility;
            //Add it to the queue
			[[PDFKThumbQueue sharedQueue] addFetchOperation:thumbFetch];
		}
		return object;
	}
}

- (void)setObject:(UIImage *)image forKey:(NSString *)key
{
	@synchronized(thumbCache)
	{
        //Add the image to the cache
		NSUInteger bytes = (image.size.width * image.size.height * 4.0f);
		[thumbCache setObject:image forKey:key cost:bytes];
	}
}

- (void)removeObjectForKey:(NSString *)key
{
	@synchronized(thumbCache)
	{
		[thumbCache removeObjectForKey:key];
	}
}

- (void)removeNullForKey:(NSString *)key
{
	@synchronized(thumbCache)
	{
        //Remove the object only if it is a NSNull object.
		id object = [thumbCache objectForKey:key];
		if ([object isMemberOfClass:[NSNull class]])
		{
			[thumbCache removeObjectForKey:key];
		}
	}
}

- (void)removeAllObjects
{
	@synchronized(thumbCache)
	{
		[thumbCache removeAllObjects];
	}
}

@end

File: Classes/Thumbs/PDFKThumbCache.swift
//
//  PDFKThumbCache.swift
//  M13PDFKit
/*
Copyright (c) 2015 Brandon McQuilkin

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

import Foundation

//The total cost of the cache. 25MB
internal let CACHE_SIZE: Int = 26214400

/**
Stores the thumbs for later reuse.
*/
internal class PDFKThumbCache: NSObject {
    
    //------------------------------------------
    /// @name Creation
    //------------------------------------------
    
    /**
    Get the shared thumb cache.
    
    @return The single cache instance.
    */
    internal class func sharedInstance() -> PDFKThumbCache {
        struct Static {
            static var onceToken: dispatch_once_t = 0
            static var instance: PDFKThumbCache? = nil
        }
        dispatch_once(&Static.onceToken) {
            Static.instance = PDFKThumbCache()
        }
        return Static.instance!
    }
    
    /**
    Get the application cache directory. This won't change, so retreive it once, and store it.
    */
    private class func appCachesPath() -> String {
        struct StaticString {
            static var onceToken: dispatch_once_t = 0
            static var instance: String? = nil
        }
        dispatch_once(&StaticString.onceToken) {
            let cachesPaths = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.CachesDirectory, NSSearchPathDomainMask.UserDomainMask, true) as! [String]
            StaticString.instance = cachesPaths.first!.stringByAppendingPathComponent("PDFKCache")
        }
        return StaticString.instance!
    }
    
    private override init() {
        super.init()
        cache.name = "PDFKThumbCache"
        cache.totalCostLimit = CACHE_SIZE
    }
    
    //------------------------------------------
    /// @name Cache Management
    //------------------------------------------
    
    /**
    Update the modification date of the thumb cache coresponding the the GUID.
    
    @param guid The GUID of the PDF document to update the cache for.
    */
    internal class func touchThumbCacheWithGUID(guid: String) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), { () -> Void in
            NSFileManager.defaultManager().setAttributes([NSFileModificationDate: NSDate()], ofItemAtPath: PDFKThumbCache.thumbCachePathForGUID(guid), error: nil)
        })
    }
    
    /**
    Creates the cache for the PDF document with the given GUID.
    
    @param guid The GUID of the PDF document to create the cache for.
    */
    internal class func createThumbCacheWithGUID(guid: String) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), { () -> Void in
            NSFileManager.defaultManager().createDirectoryAtPath(PDFKThumbCache.thumbCachePathForGUID(guid), withIntermediateDirectories: true, attributes: nil, error: nil)
        })
    }
    
    /**
    Removes the cache for the PDF document with the given GUID.
    
    @param guid The GUID of the PDF document to remove the cache of.
    */
    internal class func removeThumbCacheWithGUID(guid: String) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), { () -> Void in
            NSFileManager.defaultManager().removeItemAtPath(PDFKThumbCache.thumbCachePathForGUID(guid), error: nil)
        })
    }
    
    /**
    Deletes all caches older than the given age.
    
    @param age The time to determine what caches to expunge.
    */
    internal class func purgeThumbCachesOlderThan(age: NSTimeInterval) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), { () -> Void in
            let now: NSDate = NSDate()
            let fileManager: NSFileManager = NSFileManager.defaultManager()
            var cachePath: String = PDFKThumbCache.appCachesPath()
            
            //Get a list of all caches
            if let cachesList: [String] = fileManager.contentsOfDirectoryAtPath(PDFKThumbCache.appCachesPath(), error: nil) as? [String] {
                
                //Iterate through all the files
                for cacheName in cachesList {
                    
                    //Get cache attributes
                    if let attributes = fileManager.attributesOfItemAtPath(cachePath.stringByAppendingPathComponent(cacheName), error:nil) {
                        
                        //Get the last modification date
                        if let cacheDate = attributes[NSFileModificationDate] as? NSDate {
                            let fileAge: NSTimeInterval = now.timeIntervalSinceDate(cacheDate)
                            //If we are older, delete
                            if fileAge > age {
                                fileManager.removeItemAtPath(cachePath.stringByAppendingPathComponent(cacheName), error: nil)
                                #if DEBUG
                                println("Purged: \(cacheName)")
                                #endif
                            }
                        }
                    }
                }
            }
        })
    }
    
    /**
    Get the cache path for the PDF document with the given GUID.
    
    @param guid The guid of the PDF document to get the cache path for.
    
    @return The cache path for the PDF document with the given GUID.
    */
    internal class func thumbCachePathForGUID(guid: String) -> String {
        return PDFKThumbCache.appCachesPath().stringByAppendingPathComponent(guid)
    }
    
    //------------------------------------------
    /// @name Caching
    //------------------------------------------
    
    private let cache: NSCache = NSCache()
    
    /**
    @synchronized replacement for Swift.
    
    @param lock The object to lock on.
    @param closure The block to run.
    */
    private func synchronized(lock: AnyObject, closure: () -> ()) {
        objc_sync_enter(lock)
        closure()
        objc_sync_exit(lock)
    }
    
    /**
    Retreives or create the given thumb for the thumb request.
    
    @param request The request for a given thumb.
    @param hasPriority If set to true, the thumb will be placed next in the queue. Otherwise the thumb request will be placed at the end of the queue.
    
    @return A UIImage of the given thumb.
    */
    internal func thumbWithThumbRequest(request: PDFKThumbRequest, hasPriority: Bool) -> UIImage? {
        //We only want one "Run" of this code running at a time. If multiple threads are running this code, we could recreate a thumb twice.
        var image: UIImage?
        synchronized(cache, closure: { () -> () in
            //See if the object already exists in the cache.
            if let object: UIImage = self.cache.objectForKey(request.cacheKey) as? UIImage {
                image = object
            } else {
                //FIXME removed NSNull placeholders, may not be necessary.
                //A thumb object does not yet exist. Fetch it
                let thumbFetcher: PDFKThumbFetcher = PDFKThumbFetcher(request: request)
                //Set the priority
                thumbFetcher.queuePriority = hasPriority ? NSOperationQueuePriority.Normal : NSOperationQueuePriority.Low
                request.thumbView?.operation = thumbFetcher
                thumbFetcher.qualityOfService = NSQualityOfService.Utility
                //Add it to the queue
                PDFKThumbQueue.sharedQueue().addFetchOperation(thumbFetcher)
            }
        })
        
        return image
    }
    
    /**
    Adds the given image to the cache with the given key.
    
    @param image: The image to add to the cache.
    @param key: The key for the image.
    */
    internal func setImage(image: UIImage, forKey key: String) {
        synchronized(cache, closure: { () -> () in
            let bytes: Int = Int(image.size.width * image.size.height * 4.0)
            self.cache.setObject(image, forKey: key, cost: bytes)
        })
    }
    
    /**
    Remove the object from the cache with the given key.
    
    @param key The key of the object to remove.
    */
    internal func removeImageForKey(key: String) {
        synchronized(cache, closure: { () -> () in
            self.cache.removeObjectForKey(key)
        })
    }
    
    /**
    Remove all objects from the cache.
    */
    internal func removeAllObjects() {
        synchronized(cache, closure: { () -> () in
            self.cache.removeAllObjects()
        })
    }
}